### PR TITLE
refactor(eventual-send): staged

### PR DESCRIPTION
**Pull request by SMotaal**
This will be pushed into #1542 as separate commits.

- [x] Use `import('./types')` instead of `typedef` aliases internally
- [x] Move `HandledPromise`-specific types into `src/handled-promise.js`
- [x] Move utility types into `src/types.d.ts`

**Future Considerations**

- Use `satisfies` instead of `expectType<…>(…)`
  - Rewrite `src/index.test-d.ts` to `src/types.test.ts`
  - Consider using `tsc` instead of `tsd` in `package.json` scripts
- Cleanup `src/index.d.ts` (drop?)

### CODE REVIEW

File: packages/eventual-send/src/index.d.ts [hunk 1]
```diff
@@ -15,6 +17,20 @@ export import DataOnly = _E.DataOnly;
...
+
+export import ResolveWithPresenceOptionsBag = _HandledPromise.ResolveWithPresenceOptionsBag;
+export import HandledExecutor = _HandledPromise.HandledExecutor;
+export import Settler = _HandledPromise.Settler;
+export import HandledPromiseStaticMethods = _HandledPromise.HandledPromiseStaticMethods;
+export import HandledPromiseConstructor = _HandledPromise.HandledPromiseConstructor;
+
+export import EHandler = _HandledPromise.Handler;
 
 // Type definitions for eventual-send
```
Comment (SMotaal) [inline packages/eventual-send/src/index.d.ts]: This is no longer relevant imho:

```ts
// Type definitions for eventual-send

/**
 * @file Type definitions for @agoric/eventual-send
 *
 * Some useful background knowledge:
 *
 * `Omit<T, U>` means to return a record type `T2` which has none of the properties whose keys are part of `U`.
 * `Omit<{a: 1, b: 2, c: 3}, 'b'>` is the type `{a: 1, c: 3}`.
 *
 * `Pick<T, U>` means to return a record type `T2` which has only the properties whose keys are part of `U`.
 * `Pick<{a: 1, b: 2, c: 3}, 'b'>` is the type `{b: 2}`.
 *
 * `PromiseLike<T>` is a thenable which resolves to `T`.
 *
 * `Promise<PromiseLike<T>>` doesn't handle recursion and is distinct from `T`.
 *
 * `Unpromise<PromiseLike<T>>` strips off just one layer and is just `T`.  `Unpromise<PromiseLike<PromiseLIke<T>>` is `PromiseLike<T>`.
 *
 * `Awaited<PromiseLike<T>>` recurses, and is just `T`.
 * `Awaited<PromiseLike<PromiseLike<T>>>` is just `T` as well.
 *
 * @see {@link https://www.typescriptlang.org/docs/handbook/2/generics.html#handbook-content}
 * @see {@link https://www.typescriptlang.org/docs/handbook/2/conditional-types.html}
 */
```

- `Unpromise` is a mystery to me and is still referenced in [packages/captp/src/ts-types.d.ts](https://github.com/SMotaal/endo/blob/smotaal/eventual-send-types-refactor-staged/packages/captp/src/ts-types.d.ts#L4) from [2 years ago](https://github.com/SMotaal/endo/commit/1b767034b3056bbd73260282d42ceb94dff6c8dd#diff-cbd98e421ec3aaa96e5494b3872eff94f5ffa8a2b4b1e7a69d15d44a6796f825R4).
- The rest is purely TypeScript.

@michaelfig remove?

File: packages/eventual-send/src/index.d.ts [hunk 2]
```diff
@@ -15,6 +17,20 @@ export import DataOnly = _E.DataOnly;
...
+export import PickCallable = _E.PickCallable;
+export import RemoteKit = _E.EPromiseKit;
+
+export import ResolveWithPresenceOptionsBag = _HandledPromise.ResolveWithPresenceOptionsBag;
+export import HandledExecutor = _HandledPromise.HandledExecutor;
+export import Settler = _HandledPromise.Settler;
+export import HandledPromiseStaticMethods = _HandledPromise.HandledPromiseStaticMethods;
+export import HandledPromiseConstructor = _HandledPromise.HandledPromiseConstructor;
+
+export import EHandler = _HandledPromise.Handler;
```
Comment (SMotaal) [inline packages/eventual-send/src/index.d.ts]: Until everything checked out, I avoided changing exposed names for now.

@michaelfig do we want to rename exposed types and alter the references in other packages as a later commit in #1542?

File: packages/eventual-send/src/postponed.js [hunk 3]
```diff
@@ -1,17 +1,11 @@
...
- */
-
 /**
  * Create a simple postponedHandler that just postpones until donePostponing is
  * called.
  *
  * @param {import('./index').HandledPromiseConstructor} HandledPromise
- * @returns {[Required<EHandler<any>>, () => void]} A pair consisting of the
- * postponedHandler and donePostponing callback.
+ * @returns {[import('./index').EHandler<any>, () => void]} postponedHandler and donePostponing callback.
```
Comment (SMotaal) [inline packages/eventual-send/src/postponed.js]: @michaelfig this is a longer discussion about `typedef` aliases vs `import` references as I am doing here. I think the details need to go into a style doc somewhere to ensure everyone is on the same page.

File: packages/eventual-send/src/types.d.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+export type Callable = (...args: any[]) => any;
```
Comment (SMotaal) [inline packages/eventual-send/src/types.d.ts]: `💬 I'm so lonely out here` said the type which is only being referenced within `E.js` at this point!

@michaelfig move to `E.js`?

Comment (michaelfig) [reply packages/eventual-send/src/types.d.ts]: You can put it where you'd like, but please change to:
```suggestion
export type Callable = (...args: unknown[]) => any;
```

File: packages/eventual-send/src/E.js [hunk 5]
```diff
@@ -242,40 +245,140 @@ export default makeE;
...
+ */
+
+/**
+ * @template T
+ * @typedef {(
+ *   T extends RemotableBrand<infer L, infer R>
+ *     ? L
+ *     : Awaited<T> extends RemotableBrand<infer L, infer R>
+ *     ? L
+ *     : T extends PromiseLike<infer U> ? Awaited<T>
```
Comment (SMotaal) [inline packages/eventual-send/src/E.js:355]:
```suggestion
 *     : T extends PromiseLike<infer U>
 *     ? Awaited<T>
```

File: packages/eventual-send/src/types.test.ts
Comment (michaelfig) [inline packages/eventual-send/src/types.test.ts:1]: This is the only part of the PR I'm unsure about landing.  Does it affect tests adversely, or is it useful?  Even if it is benign, I'm concerned that its contents will drift out of date easily unless it's actually part of CI.

**Review by michaelfig — COMMENTED** [review]
This is really great!  My only hesitation is about types.test.ts.

File: packages/eventual-send/src/types.test.ts
Comment (SMotaal) [reply packages/eventual-send/src/types.test.ts:1]: Agreed, so it is here to merely bring up the idea as a simple example for discussions that would take place down the road.

**Pull request closed without merging**